Pre-Upgrade Compatibility Check on Plugins That Could Kill Joomla Upgrade

GeraintEdwards opened the pull request:
Pull Request for Issue #31042   (edited to add testing packages)

### Summary of Changes
New method in JoomlaupdateModelDefault -> getNonCorePlugins which fetches a list of plugins filtered by optional list of plugin folders.  The returning array excludes core Joomla plugins.
PreUpdate Check now fetches a list of 'system','user','authentication','actionlog' and 'twofactorauth' plugins that could, conceviably, break a Joomla upgrade if they contain code that isn't compatible with the targetted Joomla version
The PreUpdateChecker javascript now:

1. Disables the live update submit button until the update checker has finished or the user checks a checkbox to confirm they understand the risks of upgrading without performing appropriate checks/system changes
2. Checks the list of nonCorePlugins (mentioned above) to see if they are 'compatible' with the Joomla upgrade (or are part of a 'compatible' package).  Note that 'compatible' is based on results from the extension update server.
3. After the update check has run if there are no potential incompatible plugins the confirmation checkbox is removed
4. Additionally we now report results from failed requests to update servers (timeouts, 403, 500 errors etc.) separately

The messages displayed are changed as the update checker is run.

### Testing Instructions

In a J3.10 dev site install some extensions that are NOT compatible with Joomla 4 - at least one out of date version of an extension and one where there is no published J4 compatible version. Ideally you should have one with an invalid update server entry too.

I am attaching some test packages you can use to check - make sure to enable the system plugins after installing these test extensions. The components and plugins do nothing.
[testuptodatepackage.zip](https://github.com/joomla/joomla-cms/files/5421676/testuptodatepackage.zip)
[testuxpackage.zip](https://github.com/joomla/joomla-cms/files/5421677/testuxpackage.zip)
[testwuj3ej4package.zip](https://github.com/joomla/joomla-cms/files/5421678/testwuj3ej4package.zip)
[testwuj3n4package.zip](https://github.com/joomla/joomla-cms/files/5421679/testwuj3n4package.zip)
[testwuvcpackage.zip](https://github.com/joomla/joomla-cms/files/5421680/testwuvcpackage.zip)
[testxupackage.zip](https://github.com/joomla/joomla-cms/files/5421681/testxupackage.zip)

Then open Joomla Updater component and observe the pre-update checker tab and the live update tab - as the update checker is running and after it has completed

### Actual result BEFORE applying this Pull Request

You would be able to upgrade Joomla with an incompatible plugin installed that could 'brick' (i.e. kill) your Joomla installation which would involve a lot of work to resolve.

### Expected result AFTER applying this Pull Request

The modified update checker will stop you from performing the upgrade unless you have no incompatble plugins OR you click a checkbox to say you know what you are doing.

### Documentation Changes Required

Once the  PR is finalised the update checker and update process pages will need to be updated.
https://docs.joomla.org/Pre-Update_Check

Geraint2 committed:
Alternative presentation of compatibility checks for Joomla 4 from Joomla 3.10

Geraint2 committed:
Alternative presentation of compatibility checks for Joomla 4 from Joomla 3.10

Geraint2 committed:
remove accidental commit of .gitignore

Geraint2 committed:
Handle situation where we are running 3.10 with an extension what doesn't say its compatible with 3.10 - was incorrectly being reported as not supporting Joomla updater

Geraint2 committed:
Adding user guidance notes for update check

Geraint2 committed:
Merge remote-tracking branch 'remotes/upstream/3.10-dev' into compatibilitycheck

Geraint2 committed:
code style

GeraintEdwards committed:
Update administrator/components/com_joomlaupdate/controllers/update.php

Co-authored-by: Quy <quy@fluxbb.org>

GeraintEdwards committed:
Update administrator/components/com_joomlaupdate/controllers/update.php

Co-authored-by: Quy <quy@fluxbb.org>

Geraint2 committed:
Changed styling of compatibility feedback columns so that they are aligned
Make warning message display be conditional on being relevant to the displayed output.

Geraint2 committed:
Revert accidentally committed file

Geraint2 committed:
Use tabs for HTML spacing
Alphabetical language string key definition
Resolve undeclared updateInfo['current']

Geraint2 committed:
Missing ->state when checking "if ($currentCompatibilityStatus->state == 1)"

GeraintEdwards committed:
Merge branch '3.10-dev' into compatibilitycheck

zero-24 committed:
implement php cs changes

GeraintEdwards committed:
Merge pull request #1 from zero-24/310checker

php cs for the pre upgrade checker

Geraint2 committed:
Changes suggested by zero-24
* change updating icon to Joomla ajax-loader.gif
* Styling the results with margin left and right
Also corrected language type Jomla instead of Joomla

Geraint2 committed:
Revert accidentally committed file

Geraint2 committed:
More changes suggested by zero-24
* Remove double spacing after full stop in language strings

Geraint2 committed:
Changes suggested by zero-24
'is not supported' - awaiting feedback on capitalisation

Geraint2 committed:
Changes suggested by zero-24
* js file spacing

Geraint2 committed:
Changes suggested by zero-24
* Hathor template override php layout

Geraint2 committed:
Changes suggested by zero-24
* Add link to Joomla updater instructions for developer's benefit

Geraint2 committed:
Revised version to reflect some of Brian's ideas

GeraintEdwards committed:
Merge branch '3.10-dev' into compatibilitycheck

GeraintEdwards committed:
Merge branch '3.10-dev' into compatibilitycheck

Geraint2 committed:
Revised layout in traffic light colour sequence, using 'More Detail' and chevrons to indicate more detail available link and including introductory explanation

Geraint2 committed:
Revised layout in traffic light colour sequence, using 'More Detail' and chevrons to indicate more detail available link and including introductory explanation

Geraint2 committed:
Merge branch 'compatibilitycheck' of https://github.com/GeraintEdwards/joomla-cms into compatibilitycheck

Geraint2 committed:
Revised layout in traffic light colour sequence, using 'More Detail' and chevrons to indicate more detail available link and including introductory explanation

GeraintEdwards committed:
Update administrator/language/en-GB/en-GB.com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

GeraintEdwards committed:
Update media/com_joomlaupdate/js/default.js

Co-authored-by: Quy <quy@fluxbb.org>

GeraintEdwards committed:
Update media/com_joomlaupdate/js/default.js

Co-authored-by: Quy <quy@fluxbb.org>

Geraint2 committed:
Language string tweak as per bayareajenn's suggestion

Geraint2 committed:
Merge branch '3.10-dev' of https://github.com/joomla/joomla-cms into compatibilitycheck

Geraint2 committed:
First cut at flagging plugins that need attention

Geraint2 committed:
Add language string, confirm dialog and messages

Geraint2 committed:
should not have committed .gitignore

Geraint2 committed:
phpcs correction

Quy commented on administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php:
```diff
@@ -114,12 +114,80 @@
 		<tfoot>
+		<tr id="preupdateCheckWarning">
+			<td colspan="2">
+				<div class="alert ">
```
```suggestion
				<div class="alert">
```

zero-24 commented:
hmm seems I'm missing something? I have installed the demo extensions as mention above + this plugin of my own: https://github.com/zero-24/plg_system_advancedredirect/releases/tag/1.0.3 (it does not yet have a 4.x compatible version via the update.xml)

That extensions are also correctly listed here:
![image](https://user-images.githubusercontent.com/2596554/97087239-e61b8c00-1628-11eb-9d78-8456b9435be8.png)

But the Live Updater still looks the same and does not mention the plugins?
![image](https://user-images.githubusercontent.com/2596554/97087350-aef9aa80-1629-11eb-82ba-4c6105c571b5.png)

GeraintEdwards commented:
You installed the packages from above and enabled their system plugins?

zero-24 commented:
Ah i have just installed that plugins without enableing them. Once enable they show up as expected 👍

zero-24 commented on administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php:
```diff
@@ -114,12 +114,80 @@
+					</th>
+					</thead>
+				<?php
+				foreach ($this->nonCoreCriticalPlugins as $nonCoreCriticalPlugin)
```
please use

```php
					<?php foreach ($this->nonCoreCriticalPlugins as $nonCoreCriticalPlugin) : ?>
						<tr id='plg_<?php echo $nonCoreCriticalPlugin->extension_id ?>'>
						<td>
							<?php echo JText::_($nonCoreCriticalPlugin->name); ?>
						</td>
						<?php if ($nonCoreCriticalPlugin->package_id > 0) : ?>
							<?php foreach ($this->nonCoreExtensions as $nonCoreExtension) : ?>
								<?php if ($nonCoreCriticalPlugin->package_id == $nonCoreExtension->extension_id) : ?>
									<td>
										<?php echo $nonCoreExtension->name; ?>
									</td>
								<?php endif; ?>
							<?php endforeach; ?>
						<?php else : ?>
							<td/>
						<?php endif; ?>
						</tr>
					<?php endforeach; ?>
```

GeraintEdwards replied:
Personally I can't stand 

if : 
else :
endif;

to me it makes code far less easy to scan read.

But there seems to be heavy usage in Joomla so I will fall into line :(

zero-24 replied:
It is just used in layout files to make it easier to understand and change for people who want to override that view.

zero-24 commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -54,6 +55,8 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_DIRECTORY="FTP Root"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_RUNNING_PRE_UPDATE_CHECKS="Running Pre-Update Checks"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_RUNNING_PRE_UPDATE_CHECKS_NOTES="Please be patient whilst we run the pre-update checks on your extensions."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_PRE_UPDATE_CHECKS_FAILED="Pre-Update Checks Failed"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_PRE_UPDATE_CHECKS_FAILED_NOTES="It was not possible to check the compatibility of these plugins.  The request to the update server either timed out or returned an error"
```
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_PRE_UPDATE_CHECKS_FAILED_NOTES="It was impossible to check the compatibility of these plugins. The request to the update server either timed out or returned an error"
```

infograf768 replied:
rather: It was impossible etc.

zero-24 replied:
Updated thanks

brianteeman replied:
impossible is not correct. it is not the same as saying something was not possible.

zero-24 commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -73,6 +76,12 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_WARNING_UNKNOWN="Unkown error"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_CONFIRMATION="Do you wish to ignore the warnings about potentially incompatible plugins and to proceed with the upgrade"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN="Potentially Serious Problem"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_CONFIRM_MESSAGE="Are you sure you wish to proceed with the upgrade and ignore the warnings about potentially incompatible plugins?"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_DESC="The extension includes a plugin that could cause your upgrade to fail.<br><br>To perform the Joomla upgrade safely you should either upgrade this extension to a version compatible with your target version of Joomla or disable the relevant plugin(s) and check again.<br><br>For more information about the relevant plugins please check  ...
```
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_DESC="This extension includes a plugin that could cause your upgrade to fail.<br><br>To perform the Joomla upgrade safely you should either upgrade this extension to a version compatible with your target version of Joomla or disable the relevant plugin(s) and check again.<br><br>For more information about the relevant plugins please check the 'Live Update' tab."
```

zero-24 commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -73,6 +76,12 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_WARNING_UNKNOWN="Unkown error"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN="Potentially Serious Problem"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_CONFIRM_MESSAGE="Are you sure you wish to proceed with the upgrade and ignore the warnings about potentially incompatible plugins?"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_DESC="The extension includes a plugin that could cause your upgrade to fail.<br><br>To perform the Joomla upgrade safely you should either upgrade this extension to a version compatible with your target version of Joomla or disable the relevant plugin(s) and check again.<br><br>For more information about the relevant plugins please check  ...
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_LIST="These plugins you could cause problems during the upgrade"
```
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_LIST="The following plugins could cause problems during the upgrade"
```

infograf768 replied:
`you` ?

zero-24 replied:
Fixed thanks JM

zero-24 commented on media/com_joomlaupdate/js/default.js:
```diff
@@ -151,10 +172,25 @@ function extractionMethodHandler(target, prefix)
     PreUpdateChecker.setResultView = function (extensionData) {
         var html = '';
 
+        // simulate failure or timeout
```
Please remove this debug code here

zero-24 commented on administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php:
```diff
@@ -114,12 +114,80 @@
+		<tr id="preupdateCheckWarning">
+			<td colspan="2">
+				<div class="alert ">
+					<h4 class="alert-heading"><?php echo JText::_('WARNING'); ?></h4>
```
```suggestion
					<h4 class="alert-heading">
						<?php echo JText::_('WARNING'); ?>
					</h4>
```

zero-24 commented on administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php:
```diff
@@ -114,12 +114,80 @@
+				<div class="alert ">
+					<h4 class="alert-heading"><?php echo JText::_('WARNING'); ?></h4>
+					<div class="alert-message">
+						<div class="preupdateCheckIncomplete"><?php echo JText::_('COM_JOOMLAUPDATE_PREUPDATE_CHECK_NOT_COMPLETE'); ?></div>
```
```suggestion
						<div class="preupdateCheckIncomplete">
							<?php echo JText::_('COM_JOOMLAUPDATE_PREUPDATE_CHECK_NOT_COMPLETE'); ?>
						</div>
```

zero-24 commented on administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php:
```diff
@@ -114,12 +114,80 @@
+		<tr id="preupdateCheckCompleteProblems" class="hidden">
+			<td colspan="2">
+				<div class="alert ">
+					<h4 class="alert-heading"><?php echo JText::_('WARNING'); ?></h4>
```
```suggestion
					<h4 class="alert-heading">
						<?php echo JText::_('WARNING'); ?>
					</h4>
```

zero-24 commented on administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php:
```diff
@@ -114,12 +114,80 @@
+				<div class="alert ">
+					<h4 class="alert-heading"><?php echo JText::_('WARNING'); ?></h4>
+					<div class="alert-message">
+						<div class="preupdateCheckComplete"><?php echo JText::_('COM_JOOMLAUPDATE_PREUPDATE_CHECK_COMPLETED_YOU_HAVE_DANGEROUS_PLUGINS'); ?></div>
```
```suggestion
						<div class="preupdateCheckComplete">
							<?php echo JText::_('COM_JOOMLAUPDATE_PREUPDATE_CHECK_COMPLETED_YOU_HAVE_DANGEROUS_PLUGINS'); ?>
						</div>
```

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -1446,6 +1446,62 @@ private static function isNonCoreExtension($extension)
+		return $rows;
+	}
+
+
```
```suggestion
```

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -1446,6 +1446,62 @@ private static function isNonCoreExtension($extension)
+	 *
+	 * @return  array  name,version,updateserver
+	 *
+	 * @since   3.10.0
```
```suggestion
	 * @since   __DEPLOY_VERSION__
```

zero-24 commented:
@brianteeman can we please get your feedback on the new process + the texts proposed here? Thanks!

zero-24 commented:
@GeraintEdwards Thanks for that PR i think this is a good step forward.

What do you think about the idea to allow the plugins to be disable on the `Live Update` tab similiar to the RFC that Brian did?

The basic idea would be to change the table you did here:
![image](https://user-images.githubusercontent.com/2596554/97089110-e706ea80-1635-11eb-8660-e245a7e66eb1.png)

To something like this (sorry very bad paint skills on my side)
![Screenshot 2020-10-24 201736](https://user-images.githubusercontent.com/2596554/97089172-6399c900-1636-11eb-8931-ffc924102535.png)

And all plugins checked there get logged & disabled as first step here: https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_joomlaupdate/controllers/update.php#L107

GeraintEdwards commented:
The idea of disabling selected plugins is good but there are potential downsides:
- We would need to distinguish between plugins to disable temporarily (with the session, if the session data is safe??) to be re-enabled on completion and those that should be disabled permanently.
- The user may not realise that disabling a plugin could have unforseen consequences - the site may be inaccessible after the upgrade completes.
- Is it possible that disabling the plugins at the start of the upgrade process could cause the site to die?? Not sure as I've not checked which plugin triggers are fired after this point?  onAfterRender etc??

Language comments-

- Some good catches.  Thanks
- 'impossible' vs 'not possible'.  To me 'impossible' has an air of permanence where, in reality, simply trying again could resolve the issue (e.g. problem was caused a timeout).  Happy to go with the majority view though.

zero-24 commented:
> The idea of disabling selected plugins is good but there are potential downsides:
> - We would need to distinguish between plugins to disable temporarily (with the session, if the session data is safe??) to be re-enabled on completion and those that should be disabled permanently.

Anything checked here should be disabled permanently. As we can not be sure that the session is passed over.

> - The user may not realise that disabling a plugin could have unforseen consequences - the site may be inaccessible after the upgrade completes.

Well he accepted that but now the upgraded site is not bricked so he can come back and fix the issues. And I'm sure that when someone has a that critical plugin installed they check that upfront. I'm not sure whether we can do much more at this point.

> - Is it possible that disabling the plugins at the start of the upgrade process could cause the site to die?? Not sure as I've not checked which plugin triggers are fired after this point?  onAfterRender etc??

At that point where i suggested to disable the plugins that system events where executed already but all within the 3.x code so all should be fine. The problem is when they are expecuted in the middel of the upgrade process but at that time they are no longer active when checked.

> Language comments-
> 
> - Some good catches.  Thanks
> - 'impossible' vs 'not possible'.  To me 'impossible' has an air of permanence where, in reality, simply trying again could resolve the issue (e.g. problem was caused a timeout).  Happy to go with the majority view though.

Same for me. I have mention brian above so we can get his view on that strings form a native speaker perspective.

GeraintEdwards committed:
Update administrator/components/com_joomlaupdate/models/default.php

Co-authored-by: Tobias Zulauf <zero-24@users.noreply.github.com>

GeraintEdwards committed:
Update administrator/components/com_joomlaupdate/models/default.php

Co-authored-by: Tobias Zulauf <zero-24@users.noreply.github.com>

GeraintEdwards committed:
Update administrator/language/en-GB/en-GB.com_joomlaupdate.ini

Co-authored-by: Tobias Zulauf <zero-24@users.noreply.github.com>

GeraintEdwards committed:
Update administrator/language/en-GB/en-GB.com_joomlaupdate.ini

Co-authored-by: Tobias Zulauf <zero-24@users.noreply.github.com>

GeraintEdwards committed:
Update administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php

Co-authored-by: Tobias Zulauf <zero-24@users.noreply.github.com>

GeraintEdwards committed:
Update administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php

Co-authored-by: Tobias Zulauf <zero-24@users.noreply.github.com>

GeraintEdwards committed:
Update administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php

Co-authored-by: Tobias Zulauf <zero-24@users.noreply.github.com>

GeraintEdwards committed:
Update administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php

Co-authored-by: Tobias Zulauf <zero-24@users.noreply.github.com>

Geraint2 committed:
zero-24 suggestions

Geraint2 committed:
Merge branch 'compatibilitycheck' of https://github.com/GeraintEdwards/joomla-cms into compatibilitycheck

GeraintEdwards committed:
Merge branch '3.10-dev' into compatibilitycheck

brianteeman commented:
Geraint is correct.

brianteeman commented:
We need to switch the extension checks and the system checks around or move extensions to its own tab because as it is right now anyone using a screen with <900px approx may not even know there is the extension section

![image](https://user-images.githubusercontent.com/1296369/97778940-ad246f80-1b72-11eb-8300-4170ac9945fa.png)

brianteeman commented:
It was only because I was looking for some strings that I see in the code that I realised that there was a popover

![image](https://user-images.githubusercontent.com/1296369/97779030-684d0880-1b73-11eb-8d5a-93e3195b3642.png)




### Suggestion
Can we add a button please or something to make it absolutely blatantly in your face obvious. Perhaps
![image](https://user-images.githubusercontent.com/1296369/97779154-3ab48f00-1b74-11eb-8994-d7ca855d85e9.png)

brianteeman commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -29,6 +29,7 @@ COM_JOOMLAUPDATE_TOOLBAR_CHECK="Check for Updates"
 COM_JOOMLAUPDATE_PREUPDATE_HEADING_CHECKED="Checked"
 COM_JOOMLAUPDATE_PREUPDATE_HEADING_REQUIREMENT="Requirement"
 COM_JOOMLAUPDATE_PREUPDATE_UNKNOWN_EXTENSION_MANIFESTCACHE_VERSION="Unknown Version"
+COM_JOOMLAUPDATE_PREUPDATE_CHECK_NOT_COMPLETE="Pre-Update checks have not been completed yet - please wait"
```
```suggestion
COM_JOOMLAUPDATE_PREUPDATE_CHECK_NOT_COMPLETE="Pre-Update checks have not been completed yet - please wait."
```

brianteeman commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -54,6 +55,8 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_DIRECTORY="FTP Root"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_RUNNING_PRE_UPDATE_CHECKS="Running Pre-Update Checks"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_RUNNING_PRE_UPDATE_CHECKS_NOTES="Please be patient whilst we run the pre-update checks on your extensions."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_PRE_UPDATE_CHECKS_FAILED="Pre-Update Checks Failed"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_PRE_UPDATE_CHECKS_FAILED_NOTES="It was not possible to check the compatibility of these plugins.  The request to the update server either timed out or returned an error"
```
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_PRE_UPDATE_CHECKS_FAILED_NOTES="It was not possible to check the compatibility of these plugins. The request to the update server either timed out or returned an error."
```

remove double space and add full stop

GeraintEdwards replied:
I'm an old fashioned typist and old habits like 2 spaces after a full stop are difficult to kick

brianteeman commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -73,6 +76,12 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_WARNING_UNKNOWN="Unkown error"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_SERVER_ERROR="Update Server error"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS="Extensions Pre-Update Check"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_NONE="No extensions installed."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_BEING_CHECKED="The system is currently checking these plugins to see if they could cause problems during the upgrade.<br>Please be patient while the checks are completed."
```
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_BEING_CHECKED="Checking these plugins to see if they could cause problems during the upgrade.<br>Please be patient while the checks are completed."
```
Not really sure when this is displayed as I couldnt see it BUT need to be change the language or it might be interpreted as saying that joomla is checking the code in the plugins

GeraintEdwards replied:
If you switch to the 'live update' tab as soon as you load the update page then its shown next to the list of plugins.  

It only really shows up if the update checks are being slow and is mean to encourage patience!  'The system is currently checking' was meant to convey that the website is doing something right now and the user needs to wait.

brianteeman commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -73,6 +76,12 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_WARNING_UNKNOWN="Unkown error"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS="Extensions Pre-Update Check"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_NONE="No extensions installed."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_BEING_CHECKED="The system is currently checking these plugins to see if they could cause problems during the upgrade.<br>Please be patient while the checks are completed."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_CONFIRMATION="Do you wish to ignore the warnings about potentially incompatible plugins and to proceed with the upgrade"
```
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_CONFIRMATION="Do you wish to ignore the warnings about potentially incompatible plugins and to proceed with the upgrade?"
```

brianteeman commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -73,6 +76,12 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_WARNING_UNKNOWN="Unkown error"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_BEING_CHECKED="The system is currently checking these plugins to see if they could cause problems during the upgrade.<br>Please be patient while the checks are completed."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_CONFIRMATION="Do you wish to ignore the warnings about potentially incompatible plugins and to proceed with the upgrade"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN="Potentially Serious Problem"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_CONFIRM_MESSAGE="Are you sure you wish to proceed with the upgrade and ignore the warnings about potentially incompatible plugins?"
```
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_CONFIRM_MESSAGE="Are you sure you want to ignore the warnings about potentially incompatible plugins and proceed with the upgrade?"
```

puts the important part of the message first

brianteeman commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -73,6 +76,12 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_WARNING_UNKNOWN="Unkown error"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_CONFIRMATION="Do you wish to ignore the warnings about potentially incompatible plugins and to proceed with the upgrade"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN="Potentially Serious Problem"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_CONFIRM_MESSAGE="Are you sure you wish to proceed with the upgrade and ignore the warnings about potentially incompatible plugins?"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_DESC="This extension includes a plugin that could cause your upgrade to fail.<br><br>To perform the Joomla upgrade safely you should either upgrade this extension to a version compatible with your target version of Joomla or disable the relevant plugin(s) and check again.<br><br>For more information about the relevant plugins please check ...
```
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_DESC="This extension includes a plugin that could cause the upgrade to fail.<br><br>To perform the Joomla upgrade safely you should either upgrade this extension to a version compatible with your target version of Joomla or disable the relevant plugin(s) and check again.<br><br>For more information about the relevant plugins please check the 'Live Update' tab."
```

brianteeman commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -73,6 +76,12 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_WARNING_UNKNOWN="Unkown error"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_CONFIRMATION="Do you wish to ignore the warnings about potentially incompatible plugins and to proceed with the upgrade"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN="Potentially Serious Problem"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_CONFIRM_MESSAGE="Are you sure you wish to proceed with the upgrade and ignore the warnings about potentially incompatible plugins?"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_POTENTIALLY_DANGEROUS_PLUGIN_DESC="This extension includes a plugin that could cause your upgrade to fail.<br><br>To perform the Joomla upgrade safely you should either upgrade this extension to a version compatible with your target version of Joomla or disable the relevant plugin(s) and check again.<br><br>For more information about the relevant plugins please check ...
```
>> For more information about the relevant plugins please check the 'Live Update' tab.

Can we not display this on the same page?

GeraintEdwards replied:
I placed the information directly above the "submit" button to make it really obvious to the user.  The downside, as you point out, is the information in 2 different places.

brianteeman commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -100,6 +109,7 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_NO_LIVE_UPDATE_DESC="There is a new version of the
 COM_JOOMLAUPDATE_VIEW_DEFAULT_PACKAGE="Update package URL"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_PACKAGE_REINSTALL="Reinstall package URL"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_PREUPDATE_CHECK="Pre-Update Check"
+COM_JOOMLAUPDATE_PREUPDATE_CHECK_COMPLETED_YOU_HAVE_DANGEROUS_PLUGINS="You have plugins installed and enabled that could interfere with the Joomla upgrade and could result in an upgrade failure that leaves your site inaccessible.<br><br>You are strongly advised to upgrade, disable or uninstall these plugins prior to upgrading your site."
```
```suggestion
COM_JOOMLAUPDATE_PREUPDATE_CHECK_COMPLETED_YOU_HAVE_DANGEROUS_PLUGINS="There are plugins installed and enabled that could interfere with the Joomla upgrade and result in a failed upgrade that leaves the site inaccessible.<br><br>You are strongly advised to upgrade, disable or uninstall these plugins before upgrading."
```

brianteeman commented:
general comment on strings. some people might prefer to use "may" instead of "could" - i have no preference

if possible we should talk about the website and not your site

GeraintEdwards committed:
Update administrator/language/en-GB/en-GB.com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

GeraintEdwards committed:
Update administrator/language/en-GB/en-GB.com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

GeraintEdwards committed:
Update administrator/language/en-GB/en-GB.com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

GeraintEdwards committed:
Update administrator/language/en-GB/en-GB.com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

GeraintEdwards committed:
Update administrator/language/en-GB/en-GB.com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

GeraintEdwards committed:
Update administrator/language/en-GB/en-GB.com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

GeraintEdwards commented:
> We need to switch the extension checks and the system checks around or move extensions to its own tab because as it is right now anyone using a screen with <900px approx may not even know there is the extension section
re switching the ordering of the sections what about using 'More Detail' 

e.g. 
"Required PHP & Database Settings"  Passed [More Detail]
"Recommended PHP Settings" Warning [More Detail]
the warning could highlight that output buffering is enabled

brianteeman commented:
That could work as well

GeraintEdwards commented:
> It was only because I was looking for some strings that I see in the code that I realised that there was a popover
> 
> ![image](https://user-images.githubusercontent.com/1296369/97779030-684d0880-1b73-11eb-8d5a-93e3195b3642.png)
> ### Suggestion
> 
> Can we add a button please or something to make it absolutely blatantly in your face obvious. Perhaps
> ![image](https://user-images.githubusercontent.com/1296369/97779154-3ab48f00-1b74-11eb-8994-d7ca855d85e9.png)
Good idea
On a separate note I think Joomla needs a 'house style' for 'information available' or 'action required' icons/buttons etc.  As you say avilability of  popovers/tooltips is not always obvious!

Geraint2 committed:
Updates as suggested in Git Discussion to make potentially serious problem message clearer and to reduce prominence of php settings check adding in More Detail option

Geraint2 committed:
Merge remote-tracking branch 'origin/compatibilitycheck' into compatibilitycheck

# Conflicts:
#	administrator/language/en-GB/en-GB.com_joomlaupdate.ini

Geraint2 committed:
Updates as suggested in Git Discussion

GeraintEdwards commented:
@brianteeman I've updated the code now.

1. I'm not happy with the 'more information' icon/button - I think it would be cleaner to indication within the yellow warning that some more information is available  Not sure the best way for this to be clear and to look good
2. The chevron in the php settings is not rendered properly straightaway - This could be fixed by causing the preupdate checker to wait for DOMContentReady before starting.  This is probably a good idea and will only delay things by a fraction of a second.

@zero-24 and @brianteeman 

Any thoughts on the layout of the list of plugins and their packages on the 'live update' tab - we could try to present information about the author of the extension to display here and a link to the site (if available).  Anything else?

GeraintEdwards committed:
Merge branch '3.10-dev' into compatibilitycheck

zero-24 commented:
> Any thoughts on the layout of the list of plugins and their packages on the 'live update' tab - we could try to present information about the author of the extension to display here and a link to the site (if available). Anything else?

Yes I would propose to link the authorUrl / packagerUrl (for packages) with the author / packager (for packages). When there is no link than just the name, when there is no name but a link maybe fallback to "Developers website" with the URL when nothing is aviabe show that we have no information about the developer.

Geraint2 committed:
Add problem plugin author and author URL information to the warning information
Fix for javascript error thrown when no no live update is available
Start update checker on DOMContentLoaded

GeraintEdwards commented:
> > Any thoughts on the layout of the list of plugins and their packages on the 'live update' tab - we could try to present information about the author of the extension to display here and a link to the site (if available). Anything else?
> 
> Yes I would propose to link the authorUrl / packagerUrl (for packages) with the author / packager (for packages). When there is no link than just the name, when there is no name but a link maybe fallback to "Developers website" with the URL when nothing is aviabe show that we have no information about the developer.

I've done this now.  It outputs the author and the authorUrl from the plugin manifest and if not available the equivalent values from  the package manifest.

Geraint2 committed:
phpcs

adj9 commented:
Start with J!3.10
- Install all plugin 
- Enable plg

### Before apply PR
![Schermata 2020-11-14 alle 11 47 42](https://user-images.githubusercontent.com/5918444/99145686-16c37400-2671-11eb-9422-9a805fc2cbc7.png)
### After apply PR
![Schermata 2020-11-14 alle 11 58 34](https://user-images.githubusercontent.com/5918444/99145699-365a9c80-2671-11eb-94a3-bfcc42a35cda.png)
![Schermata 2020-11-14 alle 11 58 50](https://user-images.githubusercontent.com/5918444/99145702-3c507d80-2671-11eb-9ae8-aea5a4f55219.png)
The plg are confermed but there are 2 allert red & 1 orange.

particthistle commented:
@GeraintEdwards taking a look at this today following your mention in BF@H channel.

_When this is finally merged, please advise (if I don't spot it first) as I've added this to the Joomla 4 JCM series, and will also help with documenting the process on JDOCS._

I'm testing 3.10 on the skeleton build I use for implementing new sites (and my, there's a lot of extensions needing update compatibility information added) so it's got a considerable list of 405 extensions in the Manage list.

So performance issues aside, here's some feedback on what I've seen going through the comment history above.

- The "Update Required" section I think should come first in the list - Extensions identified as Update Required really should be updated first, as it's possible the update contains the Update Information, and they should be ideally updated regardless of what the Update Information availability is returned as).

- [More Detail >] was previously positioned to the right which I think looks better. Otherwise, it should go under the section title if not back on the right.

- With 405 extensions checked (less core) the groupings are quite lengthy. Would it be better to put each section in an accordion slider, so that then the results are presented in groups with the description for that section visible, and then an expand button to then open up the list?

- Is hiding the information that is displayed when you click on [More Detail >] worth hiding (is there history behind it?)

- Should the columns read left to right: Extension Name, Installed Version, Extension Type, Update Compatible Version, Currently Compatible Version, Update Compatible Version

- "Update Compatible Version" and "Currently Compatible Version" - compatible to what? Do we need to perhaps make these language strings more specific if this is what they actually refer to?
Currently Compatible Version = Version compatible with Joomla 3.x
Update Compatible Version = Version compatible with Joomla 4.x

On applying the current version of the patch, but before trying to install a test extension:
- the [More Detail >] links are no longer clickable.
- the Extensions Pre-Update Check did not automatically start when the patch was applied

I've then installed the test extensions, but as the pre-update check did not trigger, and could not be manually triggered, I couldn't see the results for the 6 test packages yet.

Geraint2 committed:
Fix for preupdatechecker not auto-starting automatically

GeraintEdwards commented:
> Start with J!3.10 alpha4-dev
> ...
> The plg are confermed but there are 2 allert red & 1 orange.

Somehow the auto-running of the PreUpdateChecker was broken in my last commit :(  This is now working

GeraintEdwards commented:
Thanks @particthistle 
 
>     * The "Update Required" section I think should come first in the list - Extensions identified as Update Required really should be updated first, as it's possible the update contains the Update Information, and they should be ideally updated regardless of what the Update Information availability is returned as).
 
>     * [More Detail >] was previously positioned to the right which I think looks better. Otherwise, it should go under the section title if not back on the right.
I had broken the auto-running of the checker - I have corrected this now and the [More Detail >] is back where is should be!
 
>     * With 405 extensions checked (less core) the groupings are quite lengthy. Would it be better to put each section in an accordion slider, so that then the results are presented in groups with the description for that section visible, and then an expand button to then open up the list?

I'm open to this idea or a toggle to minimise each block but I think they should be fully visible initially·   What do you think?

>     * Is hiding the information that is displayed when you click on [More Detail >] worth hiding (is there history behind it?)

My initial thought was to have it all visible but it was suggested this could make it overwhelming to some users.

Is there a way to do a vote on this type of question in GitHub??

>     * Should the columns read left to right: Extension Name, Installed Version, Extension Type, Update Compatible Version, Currently Compatible Version, Update Compatible Version

Or 
"Extension Name, Extension Type, Installed Version, Update Compatible Version, Currently Compatible Version" ?

Again how do we ascertain consensus?

>     * "Update Compatible Version" and "Currently Compatible Version" - compatible to what? Do we need to perhaps make these language strings more specific if this is what they actually refer to?
>       Currently Compatible Version = Version compatible with Joomla 3.x
>       Update Compatible Version = Version compatible with Joomla 4.x

A good idea - but we need concise headings too.  So 'Joomla 3.x Compatible Version' etc.

The number values should not be hard coded as 4.x etc. since the check will work with whatever the target Joomla version is e.g. "4.0.0 beta 5" of "4.1.0 RC2" etc.

The descriptors then get a bit long

richard67 commented:
@zero-24 How do we continue with this PR? It shows that you have requested changes during a review. I've gone through the history to see if that meanwhile has been resolved, but I could not really find the details. Could you check and if ok just give again a review with approval so the change request will disappear at the bottom of the PR?

zero-24 commented:
I would say you can remove my review given that they are indeed solved. The next steps here would be testing i would say.

richard67 committed:
Merge branch '3.10-dev' into compatibilitycheck

richard67 commented:
@adj9 @particthistle Could you test this PR again after it has received changes? Thanks in advance.

GeraintEdwards commented:
In addition to testing @particthistle bumping questions from https://github.com/joomla/joomla-cms/pull/31200#issuecomment-727225267 please

GeraintEdwards committed:
Merge branch '3.10-dev' into compatibilitycheck

zero-24 committed:
Merge branch '3.10-dev' into compatibilitycheck

zero-24 commented:
I'm sorry that it took that long to get back into this. I have just tested the process and I think this is a good step in the right direction.

@GeraintEdwards 
While i was testing i have "just" installed the extensions you have provided here in the inital issue but i have not enabled any additional extension. While at this point no "brick my site" plugin was active it seems the headers for the table was still rendered.  I would say lets just remove the table in that case.

![image](https://user-images.githubusercontent.com/2596554/104975075-ef36b200-59f9-11eb-8440-5489dd17c745.png)

The other case I have tested is an 3.10 without any extension installed:
![image](https://user-images.githubusercontent.com/2596554/104974847-6a4b9880-59f9-11eb-8301-083166040365.png)

The `Live Update` screen claims to still checking while the pre Upgrade checker already issued that no extension is installed. So in this case i would propose to remove the warning as well as the table header etc too.

A tip for all testers go to the disabled extensions and enabled them to get plugins that are not compatible.

GeraintEdwards committed:
Merge branch '3.10-dev' into compatibilitycheck

Geraint2 committed:
Change language string  COM_JOOMLAUPDATE_VIEW_DEFAULT_SHOW_MORE_COMPATIBILITY_INFORMATION to COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_SHOW_MORE_COMPATIBILITY_INFORMATION so that it fits the relevant pattern of strings better
Change "Currently Compatible Version" and "Update Compatible Version" to refer to current and target Joomla versions (@particthistle)
Re-ordered columns -> "Extension Name, Extension Type, Installed Version, Joomla '3.x' Compatible Version, Joomla '4.x' Compatible Version" to make it easier to read and show progression of Joomla versions (@particthistle)
Hide plugin warning and table headings as soon as all the plugins have been tested if they are all 'safe' on the 'live update' tab (@zero-24)

GeraintEdwards commented:
@zero-24 I have updated the JS to hide the table headings when there are no problem plugins and to do so as soon as the list has been cleared (as opposed to waiting for the full checks to complete)

@particthistle I have re-ordered the output columns and updated the labels to show the current Joomla version and target version which should make them easier to interpret.

I think this PR is ready for final testing now.

zero-24 commented:
@GeraintEdwards the first issue mentind is fixed now Thanks! But the seccond one is still there.

When you have zero extensions installed the `live update` claims to `Pre-Update checks have not been completed yet - please wait.` But when checking the `Pre-Update check` it correctly states `No extensions installed.`.

![image](https://user-images.githubusercontent.com/2596554/106518187-d0481d80-64d9-11eb-8bd9-fb4d4d054877.png)

![image](https://user-images.githubusercontent.com/2596554/106518229-dc33df80-64d9-11eb-8fef-8c432bbbbad6.png)

Geraint2 committed:
When there are no extensions installed or no non-Core Critical Plugins installed disable the wanrning messages and immediately exit the checking process.

GeraintEdwards commented:
Good catch on the no-extensions and no non-core critical plugins installed scenarios.  I have amended the javascript to reflect these situations now.

GeraintEdwards committed:
Merge branch '3.10-dev' into compatibilitycheck

GeraintEdwards committed:
Merge branch '3.10-dev' into compatibilitycheck

alikon commented:
I have tested this item :white_check_mark: successfully on b3e691aebdb5d6b4f288aadef6aece6b0cbc356a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31200">issues.joomla.org/tracker/joomla-cms/31200</a>.</sub>

particthistle commented:
I have tested this item :red_circle: unsuccessfully on b3e691aebdb5d6b4f288aadef6aece6b0cbc356a<br /><br />Sorry @GeraintEdwards - will chat more at JUGL on this. Apologies for delay looking at it after changes back in December. I've now adjusted my Github notifications so they appear somewhere I see them when I'm not on Github.

Testing unsuccessfully based on the UX more than the functionality (also reviewing [my comment from last year](#event-569289)).

The "check" process appears to work fine, albeit that the bulk of extensions returning Update Information Unavailable (@Llewellynvdm - a discussion JED need to have with developers to onboard them to this process)

I've recorded a screencast to show where the bugs lie: https://www.screencast.com/t/VygEBpCvR721
- Accordions for information blocks are not toggling (tested in both Firefox and Chrome on Windows)
- Language strings in the process do not indicate anywhere that the Pre-Update Check is actually checking everything in preparation of installing Joomla 4.
- Link to JDOCs page on "pre-update check documentation" should open in a new tab/window.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31200">issues.joomla.org/tracker/joomla-cms/31200</a>.</sub>

particthistle commented:
The issue with the toggles looks to be string that's not defined (my JQuery skills are not good enough to suggest a fix).

![image](https://user-images.githubusercontent.com/5515866/111244721-93825100-8657-11eb-8614-f4af3c064fcf.png)

zero-24 commented:
> Language strings in the process do not indicate anywhere that the Pre-Update Check is actually checking everything in preparation of installing Joomla 4.

That has been patched outside of this PR: https://github.com/joomla/joomla-cms/pull/32098

Given that the pre Upgrade checker does not always check against Joomla 4 but the next target version provided by the updater.

richard67 commented:
> > Language strings in the process do not indicate anywhere that the Pre-Update Check is actually checking everything in preparation of installing Joomla 4.
> 
> That has been patched outside of this PR: #32098

Then it would maybe be good to update this PR to latest 3.10-dev so it contains this change, i.e. merge the 3.10-dev branch from upstream into the branch for this PR here.

@GeraintEdwards Let us know when you need help with that.

GeraintEdwards commented:
@particthistle 
> The issue with the toggles looks to be string that's not defined (my JQuery skills are not good enough to suggest a fix).> 
> ![image](https://user-images.githubusercontent.com/5515866/111244721-93825100-8657-11eb-8614-f4af3c064fcf.png)
This is a strange error as these variables are declared in default_preupdatecheck.php and I don't see this error arising.  Viewing the web page source I can see them defined.

The only 2 possible explanations I can think of for you seeing these problems is that:
1. You have some sort or plugin that defers the execution of 'inline' javascript code so that the update check starts before these variables are parsed ??
2. The sheer number of extensions you have is causing an issue in some way

GeraintEdwards commented:
@richard67 
> > > Language strings in the process do not indicate anywhere that the Pre-Update Check is actually checking everything in preparation of installing Joomla 4.
> > 
> > 
> > That has been patched outside of this PR: #32098

This has been merged so I see the changes here too.

chmst commented:
I have tested this item :white_check_mark: successfully on b3e691aebdb5d6b4f288aadef6aece6b0cbc356a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31200">issues.joomla.org/tracker/joomla-cms/31200</a>.</sub>

alikon commented:
from my understanding this can go RTC
up to RL now :smile:

alikon commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31200">issues.joomla.org/tracker/joomla-cms/31200</a>.</sub>

zero-24 commented:
I'm now finally going to merge this here into 3.10. I would like to issue a massive thanks to @GeraintEdwards who did a great job with this PR and all the testers who tested and gave feedback.

Just to make things clear. The Pre Upgrade Checker itself is one of the main features of 3.10 and this here is an great extension to the existing tool. I'm happy to hear, read, review and accept additional improvments on this tool. So I would like to encurage everyone who has additional feedback on this tool please step up and lets discuss now what can be improved and how. Feel free to report issues or provide PRs here. But please feel also free to contact me via mail (tobias.zulauf[at]community.joomla.org) or Glip. Thanks!

richard67 commented:
@GeraintEdwards @zero-24 Is it by purpose that the minified js is deleted by this PR now in the 3.10-dev branch?

![2021-03-27_01](https://user-images.githubusercontent.com/7413183/112718374-cb5c8480-8ef2-11eb-86d8-b4b30a8c2341.png)

richard67 commented:
See PR #32883 .

zero-24 commented:
Will take a closer look over the weekend

richard67 commented:
@zero-24 Not sure what closer look it needs. If you check the folder where the missing minified script which will be added with my PR #32883 is located, you will see that for the other js files minified versions are there, only not for that one handled with my PR and mentioned in my above comment.

zero-24 commented:
Yes but when its about JS i always have to take a soccond look when at my desk, because its JS ;-)

GeraintEdwards commented:
I can confirm this minified file was accidentally removed #6c00eec7a69df98a730ea71f17d891fb33f5533e on 31/01/2021

The issue with developing/debugging JS in Joomla is that you can't debug unless you enable Joomla debug mode and then Joomla forces a new version of the file on the browser each time you refresh the page (through the 'getMediaVersion query string) so all your JS breakpoints are lost.  An easy workaround for this is to temporarily rename the min.js file .  I must have accidentally let it slip through the commit :(

zero-24 commented:
The docs page has been updated: https://docs.joomla.org/Pre-Update_Check
